### ISSUE TITLE
Audit block transforms

### ISSUE THREAD

**Issue by ellatrix** [WordPress/gutenberg#63635]
Let's audit:

* Missing block transforms.
* Existing transforms that don't make sense or behave strangely.

## Existing

* Multiple => Headings does not make sense?
* Multiple => Columns: does it make sense after a certain number? Which number?
* Does SINGLE paragraph => columns make sense?
* Should there be a quote transform for blocks nested in a quote?
* There is no way to unquote, see also #51934. Similarly there's no way to ungroup from the transforms menu.
* Group transforms: we need consistency:
  * quote => group: transforms (see also #58042)
  * cover => group: groups
  * group => cover: leaves group
  * What is the right behaviour? Transform? Or group?
  * Maybe we need two "transform" categories: transform and wrap.
* Does paragraph <> code make sense? Also maybe remove Preformatted?
* Should we remove the multi <> pull quote transform (see #11610)
* Transforming Image & Text => Image loses text.
* Image & Text => video does not make sense?
* Heading => list does not make sense?
* List => heading does not make sense either.

## Missing

* #55602
* Verse <> Preformatted
* There's no transformations if you select list items (without the parent list block)

Cc @richtabor maybe you have a good sense of what else is missing.

**Comment by cbirdsong** [comment]
I couldn't find an existing issue for this, but a paragraph block containing only an inline image should be able to be transformed into an image block.

**Comment by ellatrix** [comment]
Sure, we can add that

**Comment by richtabor** [comment]
### General notes

> Does SINGLE paragraph => columns make sense?

I think it's strange to promote **transforming to columns, unless I have multi-selected blocks**. Otherwise I just place the current block in one column, which is not quite as meaningful (and perhaps disorienting). 

> Maybe we need two "transform" categories: transform and wrap.

I'm not bullish on **transforming to "Group" from the transform panel**. It's great in the block options, as it's an action I'm taking, rather than a true transforming of the current block. We're mixing the ideas a bit including transform to group in the transforms panel. 

I don't think we need two groups. I'm not sure that wrapping to anything other than group is super helpful in the editing flow (other than the case of multi-select with columns). 

### Transform priority
The notion of transform priority is an important one. I know we have a few in there, but it would be great if we could set prioritized transforms—especially for generic content blocks—like the image block to prioritize gallery and cover for example (and review the current prioritized transforms). 

Apart from the list you made, just making Columns multi-select only, and removing "Group" as an ever-present transform (relying on the block options group functionality makes the transform panel much more intentional. Then with just a few prioritized transforms for content and images, I think it'll be super solid.

**Comment by richtabor** [comment]
> Image & Text => video does not make sense?

Agreed 

> Heading => list does not make sense?
> List => heading does not make sense either.

Yea, probably doesn't make much sense. Perhaps if you have multiple content blocks selected, you can transform to List. 

> Does paragraph <> code make sense? Also maybe remove Preformatted?

I can see these being ok perhaps. It's likely confusing for many users to see "Peformatted" though

> Multiple => Headings does not make sense?

It's probably fine.

**Comment by annezazu** [comment]
Putting out there to include a way to transform between List types if possible while I know this broader issue to allow more control ideally would need to be done: https://github.com/WordPress/gutenberg/issues/51480

For now, there's not a way to transform between ordered and unordered items (you can select the parent list and change but still, it's an extra step):

https://github.com/user-attachments/assets/2361f944-d0af-46ab-9856-bd41567ce3db

Happy to open a specific issue for this but it felt on the edge to me.

**Comment by ellatrix** [comment]
@annezazu Yes, it's a general problem with list item transforms. I was also trying to convert a list to paragraphs and couldn't after selecting multiple list items. I wonder how we could best solve this. Technically, the list items are selected, not the whole list, so no transforms match (even if we add them, you can't insert paragraphs into a list block).

* Should we add a link to select the parent block if there's no transforms? I don't really like this.
* Should we show the parent transforms here? It could be strange when not all items are selected. Would it be expected that all items transform?
* Should we attempt anyway to add list item transforms, and find a way around the limitation? This would mean detecting that the result of the transform can't be inserted (into the list), splitting the list, etc.

Anyone who cares about list that is good to ping? I think @jasmussen cares about lists and transforms 🙂

**Comment by jasmussen** [comment]
Well, specifically for lists, I'm not sure _transforms_ is the best interface at all. _It's a list block_, it's not an _ordered list block_ with a _unordered list block_ to transform into.

To that end, I'd think this is rather a challenge with the block toolbar itself, which for the List specifically could benefit from showing some of the parent properties.

There's something we could try; it might need refinement, but it still seems made for this. For the Buttons block, we have a feature to absorb into the child block the controls for the parent block. Not here how the justification controls are available regardless of which "layer" I've chosen:

![buttons](https://github.com/user-attachments/assets/ada67e13-4975-46ce-9596-f6bf0249d797)

This is not yet applied to the list block:

![list](https://github.com/user-attachments/assets/1acf8737-1b66-4c9c-a79f-0e2f5849f302)

But it seems worth trying. The only downside, the list block toolbar could get pretty long. Mockup:

<img width="756" alt="List block with absorbed control" src="https://github.com/user-attachments/assets/305e3d95-1d56-42fa-a533-143e8740b081">

**Comment by Marc-pi** [comment]
a frequent need is to convert from paragraph (link) to Buttons https://github.com/WordPress/gutenberg/issues/66846

**Comment by EvanLTD** [comment]
**Columns => Group** is currently useless. It's available in the transform menu, but doesn't actually "transform" anything. It just wraps the existing columns block inside a brand new group (if that's what we wanted, we could already accomplish it via regular grouping functionality). 

Logically, **Columns => Group** should remove all columns + column wrappers, leaving a simple group containing the inner content blocks.

**Comment by mtias** [comment]
@ellatrix should this one be closed? Not sure if it's useful anymore

**Comment by cbirdsong** [comment]
> I couldn't find an existing issue for this, but a paragraph block containing only an inline image should be able to be transformed into an image block.

> **Columns => Group** is currently useless. It's available in the transform menu, but doesn't actually "transform" anything. It just wraps the existing columns block inside a brand new group (if that's what we wanted, we could already accomplish it via regular grouping functionality).
> 
> Logically, **Columns => Group** should remove all columns + column wrappers, leaving a simple group containing the inner content blocks.

These are both definitely still issues. 

I would also call out transforming Media + Text or Cover into an Image - any text content is just dropped:

https://github.com/user-attachments/assets/ec4702fc-5eb9-429c-a324-dbc0c156c5f7

It looks like there have been a few PRs attempting to address that, but they haven't been merged:

- #70220
- #64540
